econml/tests/test_rscorer.py

```diff
@@ -20,52 +20,109 @@ def _fit_model(name, model, Y, T, X):
+                            ('sldml', SparseLinearDML(model_y=reg(), model_t=clf(), discrete_treatment=True,
+                                                        featurizer=PolynomialFeatures(degree=2, include_bias=False),
+                                                        cv=3)),
+                            ('xlearner', XLearner(models=reg(), cate_models=reg(), propensity_model=clf())),
+                            ('dalearner', DomainAdaptationLearner(models=reg(), final_models=reg(),
+                                                                  propensity_model=clf())),
+                            ('slearner', SLearner(overall_model=reg())),
+                            ('tlearner', TLearner(models=reg())),
+                            ('drlearner', DRLearner(model_propensity=clf(), model_regression=reg(),
+                                                    model_final=reg(), cv=3)),
+                            ('rlearner', NonParamDML(model_y=reg(), model_t=clf(), model_final=reg(),
+                                                    discrete_treatment=True, cv=3)),
+                            ('dml3dlasso', DML(model_y=reg(), model_t=clf(), model_final=reg(),
+                                               discrete_treatment=True, featurizer=PolynomialFeatures(degree=3), cv=3))
+                            ]
+
+                (X_train, X_val, T_train, T_val,
+                Y_train, Y_val, _, true_eff_val) = train_test_split(X, T, y, true_eff, test_size=.4)
+
+                models = Parallel(n_jobs=1, verbose=1)(delayed(_fit_model)(name, mdl,
+                                                                        Y_train, T_train, X_train)
+                                                    for name, mdl in models)
+
+                if discrete_outcome:
+                    scorer = RScorer(model_y=clf(), model_t=clf(),
+                                    discrete_treatment=True, discrete_outcome=discrete_outcome,
+                                    cv=3, mc_iters=2, mc_agg='median')
+                else:
+                    scorer = RScorer(model_y=reg(), model_t=clf(),
+                                    discrete_treatment=True, cv=3,
+                                    mc_iters=2, mc_agg='median')
+
+                scorer.fit(Y_val, T_val, X=X_val)
+                rscore = [scorer.score(mdl) for _, mdl in models]
+                rootpehe_score = [np.sqrt(np.mean((true_eff_val.flatten() - mdl.effect(X_val).flatten())**2))
+                                for _, mdl in models]
+                # Checking neg corr between rscore and rootpehe (precision in estimating heterogeneous effects)
+                assert LinearRegression().fit(np.array(rscore).reshape(-1, 1), np.array(rootpehe_score)).coef_ < 0.5
+
+                if discrete_outcome:
+                    # Removing SLearner. Only utilized for introducting variation in r- & rootpehe score

```

Comment:
I see that if you leave in SLearner in with the current code, the test does fail, but I think we can make a different change such that we can remove this if statement.

The ensemble tests fail when SLearner is included and `discrete_outcome=True` because there ends up being an inconsistency between the output shape of `const_marginal_effect` between all of the other discrete outcome models and `SLearner`. In short, the reason for this is that when `discrete_outcome=True`. the shape of y in supported models internally becomes (n, 1). However, since metalearners (and thus SLearner) don't support discrete_outcome, the shape of y internally still remains (n,). So we end up getting different const_marginal_effect shapes which makes numpy complain when e.g. taking the average over these const_marginal effects.

There are a few ways around this.
1. We can include the if-statement to exclude SLearner as you do currently.
2. We can revise the _get_data to make `y` and `T` of shape (n ,1) rather than (n,). Testing locally, this works.
3. A more appropriate long term fix might be to extend discrete_outcome functionality to metalearners, but I think that is out-of-scope for this PR.

Can we implement the 2nd workaround? I'll add a code suggestion shortly

Reply:
Yes, I definitely agree option 2 makes more sense in this scenario. Thanks for the recommendation here!